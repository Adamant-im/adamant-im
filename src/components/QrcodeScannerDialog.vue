<template>
  <v-dialog
    v-model="show"
    width="500"
  >
    <v-card>

      <!-- Camera Waiting -->
      <v-layout
        v-if="cameraStatus === 'waiting'"
        justify-center
        align-center
        class="pa-5"
      >
        <div class="title">{{ $t('scan.waiting_camera') }}</div>
        <v-progress-circular
          indeterminate
          color="primary"
          size="32"
          class="ml-3"
        />
      </v-layout>

      <!-- Camera Active -->
      <v-layout
        v-show="cameraStatus === 'active'"
        row
        wrap
      >
        <v-flex xs12>
          <video ref="camera" class="camera"></video>
        </v-flex>
        <v-flex xs12 class="pa-4">
          <h3 class="subheading text-xs-center">
            {{ $t('scan.hold_your_device') }}
          </h3>
        </v-flex>
      </v-layout>

      <!-- No Camera -->
      <v-layout
        v-if="cameraStatus === 'nocamera'"
        row
        wrap
        justify-center
        align-center
        class="text-xs-center pa-5"
      >
        <v-flex xs12>
          <h3 class="headline">{{ $t('scan.no_camera_found') }}</h3>
          <p class="mt-1 mb-0">{{ $t('scan.connect_camera') }}</p>
        </v-flex>
      </v-layout>

      <v-divider></v-divider>

      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn
          flat
          @click="show = false"
        >
          {{ $t('scan.close_button') }}
        </v-btn>
      </v-card-actions>

    </v-card>
  </v-dialog>
</template>

<script>
// @todo [Deprecation] URL.createObjectURL with media streams is deprecated and will be removed in M71, around December 2018. Please use HTMLMediaElement.srcObject instead. See https://www.chromestatus.com/features/5618491470118912 for more details.
export default {
  mounted () {
    this.init()
  },
  beforeDestroy () {
    this.destroyScanner()
  },
  computed: {
    show: {
      get () {
        return this.value
      },
      set (value) {
        this.$emit('input', value)
      }
    }
  },
  watch: {
    cameras (cameras) {
      if (cameras.length > 0) {
        this.currentCamera = this.cameras[0]
        this.scanner.start(this.currentCamera)

        this.cameraStatus = 'active'
      } else {
        this.cameraStatus = 'nocamera'
      }
    }
  },
  data: () => ({
    cameraStatus: 'waiting', // can be: waiting, active, nocamera
    Instascan: null, // ref
    scanner: null,
    currentCamera: null,
    cameras: []
  }),
  methods: {
    init () {
      return this.initScanner()
        .then(() => {
          return this.getCameras()
        })
        .catch(err => {
          this.cameraStatus = 'nocamera'
          this.$store.dispatch('snackbar/show', {
            message: this.$t('scan.something_wrong')
          })
          console.error(err)
        })
    },
    async initScanner () {
      const instascanModule = import(/* webpackMode: "lazy" */ 'instascan')

      // save reference
      this.Instascan = await instascanModule

      this.scanner = new this.Instascan.Scanner({
        video: this.$refs.camera,
        scanPeriod: 1,
        mirror: false
      })

      // attach scan event
      this.scanner.addListener('scan', (content) => {
        this.onScan(content)
      })
    },
    destroyScanner () {
      // First check if the scanner was initialized.
      // Needed when an unexpected error occurred,
      // or when the dialog closes before initialization.
      return this.scanner && this.scanner.stop()
    },
    async getCameras () {
      this.cameras = await this.Instascan.Camera.getCameras()
    },
    onScan (content) {
      this.$emit('scan', content)
      this.destroyScanner()
      this.show = false
    }
  },
  props: {
    value: {
      type: Boolean,
      required: true
    }
  }
}
</script>

<style scoped>
.camera {
  width: 100%;
  height: 300px;
  background-color: #000;
}
</style>
