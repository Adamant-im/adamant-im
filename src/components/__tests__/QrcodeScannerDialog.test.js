import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'

import mockupI18n from './__mocks__/plugins/i18n'
import mockupSnackbar from './__mocks__/store/modules/snackbar'
import QrcodeScannerDialog from '@/components/QrcodeScannerDialog'

Vue.use(Vuex)
Vue.use(VueI18n)
Vue.use(Vuetify)

/** Mock Instascan **/
jest.mock('instascan', () => {})

/**
 * Mockup store helper.
 */
function mockupStore () {
  const snackbar = mockupSnackbar()

  const store = new Vuex.Store({
    modules: {
      snackbar
    }
  })

  return {
    store,
    snackbar
  }
}

describe('QrcodeScannerDialog.vue', () => {
  let store = null
  let snackbar = null
  let i18n = null

  beforeEach(() => {
    // mockup Store
    const mockup = mockupStore()
    store = mockup.store
    snackbar = mockup.snackbar // used as reference

    // mockup i18n
    i18n = mockupI18n()
  })

  it('renders the correct markup', () => {
    const wrapper = shallowMount(QrcodeScannerDialog, {
      store,
      i18n,
      propsData: {
        value: true
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should display dialog when :value = true', () => {
    const wrapper = shallowMount(QrcodeScannerDialog, {
      store,
      i18n,
      propsData: {
        value: true
      }
    })

    // show dialog
    expect(wrapper.vm.show).toBe(true)

    // hide dialog
    wrapper.setProps({ value: false })
    expect(wrapper.vm.show).toBe(false)
  })

  // @todo mockup Instascan
})
