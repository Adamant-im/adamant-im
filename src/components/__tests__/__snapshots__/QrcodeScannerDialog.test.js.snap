// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QrcodeScannerDialog.vue renders the correct markup 1`] = `
<v-dialog-stub
  maxwidth="none"
  origin="center center"
  transition="dialog-transition"
  width="500"
>
  <v-card-stub
    elevation="2"
  >
    <v-layout-stub
      align-center=""
      class="pa-5"
      justify-center=""
      tag="div"
    >
      <div
        class="title"
      >
        Waiting camera
      </div>
       
      <v-progress-circular-stub
        class="ml-3"
        color="primary"
        indeterminate="true"
        rotate="0"
        size="32"
        value="0"
        width="4"
      />
    </v-layout-stub>
     
    <v-layout-stub
      row=""
      tag="div"
      wrap=""
    >
      <v-flex-stub
        tag="div"
        xs12=""
      >
        <video
          class="camera"
        />
      </v-flex-stub>
       
      <v-flex-stub
        class="pa-4"
        tag="div"
        xs12=""
      >
        <h3
          class="subheading text-xs-center"
        >
          
          Hold your device steady for 2-3 seconds towards the QR code you want to scan
        
        </h3>
      </v-flex-stub>
    </v-layout-stub>
     
    <!---->
     
    <v-divider-stub />
     
    <v-card-actions-stub>
      <v-spacer-stub />
       
      <v-btn-stub
        activeclass="v-btn--active"
        flat="true"
        tag="button"
        type="button"
      >
        
        Close
      
      </v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-dialog-stub>
`;
